pyccl/base.py

```diff
@@ -298,3 +299,332 @@ def increment(self):
+    representation string. They also inherit ``__eq__`` which checks for
+    representation equivalence.
+
+    In the implemented scheme, each ``CCLObject`` may have its own, specialized
+    ``__repr__`` method overloaded. Object representations have to be unique
+    for equivalent objects. If no ``__repr__`` is provided, the default from
+    ``object`` is used.
+
+    Mutation
+    --------
+    ``CCLObjects`` are by default immutable. This aims to provide a failsafe
+    mechanism, where, changing attributes has to trigger a re-computation
+    of something else inside of the instance, rather than simply doing a value
+    change.
+
+    This immutability mechanism can be safely bypassed if a subclass defines an
+    ``update_parameters`` method. ``CCLObjects`` temporarily unlock whenever
+    this method is called.
+
+    Internal State vs. Mutation
+    ---------------------------
+    Other methods that use ``setattr`` can only do that if they are decorated
+    with ``@unlock_instance`` or if the particular code block that makes the
+    change is enclosed within the ``UnlockInstance`` context manager.
+    If neither is provided, an exception is raised.
+
+    If such methods only change the instance's internal state, the decorator
+    may be called with ``@unlock_instance(mutate=False)`` (or equivalently
+    for the context manager ``UnlockInstance(..., mutate=False)``). Otherwise,
+    the instance is assumed to have mutated.
+    """
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+
+        # 1. Store the signature of the constructor on import.
+        cls.__signature__ = signature(cls.__init__)
+
+        # 2. Assign the arguments of the constructor if needed.
+        if (hasattr(cls, "__init_attrs__")
+                and cls.__init_attrs__
+                and "__init__" in vars(cls)):
+            cls.__init__ = auto_assign(cls.__init__, sig=cls.__signature__)
+
+        # 3. Replace repr (if implemented) with its cached version.
+        if "__repr__" in vars(cls):
+            # If the class defines a custom `__repr__`, this will be the new
+            # `_repr` (which is cached). Decorator `cached_property` requires
+            # that `__set_name__` is called on it.
+            bmethod = functools.cached_property(cls.__repr__)
+            cls._repr = bmethod
+            bmethod.__set_name__(cls, "_repr")
+            # Fall back to using `__ccl_repr__` from `CCLObject`.
+            cls.__repr__ = cls.__ccl_repr__
+
+        # 4. Unlock instance on specific methods.
+        def Funlock(cls, name, mutate):
+            # Allow instance to change or mutate if method `name` is called.
+            func = vars(cls).get(name)
+            if func is not None:
+                newfunc = unlock_instance(mutate=mutate)(func)
+                setattr(cls, name, newfunc)
+
+        Funlock(cls, "__init__", False)
+        Funlock(cls, "update_parameters", True)
```

Comment:
So this unlocks them permanently or just instantaneously?

Reply:
These two lines basically say the following:
*If `__init__` and/or `update_parameters` exist, unlock the object before the functions are run, and lock them again on exit.*